src/main/java/net/starlark/java/eval/StarlarkInt.java

```diff
@@ -512,10 +514,59 @@ public static StarlarkInt multiply(StarlarkInt x, StarlarkInt y) {
       return StarlarkInt.of(xl * yl);
     }
 
+    try {
+      long xl = x.toLongFast();
+      long yl = y.toLongFast();
+
+      // see Hacker's Delight, 8-2
+
+      long u0 = xl & 0xFFFFFFFFL;
```

Comment:
Let's use x, y, and z for the operands and result, and lo/hi for the names of arms. (x1 is too close to xl.)

```
// 128-bit signed multiplication, Hacker's Delight 8-2.
long xlo = xl & 0xFFFFFFFFL;
long xhi = xl >> 32;
long ylo = yl & 0xFFFFFFFFL;
long yhi = yl >> 32;
long zlo = xlo * ylo;
long t = xhi * ylo + (zlo >>> 32);
long z1 = t & 0xFFFFFFFFL;
long z2 = t >>> 32;
z1 += xlo * yhi;
long zhi = xhi * yhi + z2 + (z1 >> 32);
```

(results are zlo, zhi)

I notice that the you (and the JDK) use a logical shift (t >>> 32) whereas HD uses an arithmetic shift (t >> 32). Why is that safe? I can't prove that t is nonnegative, though I can't yet find a bug.

Is the int32 * int32 special case still warranted? It might be faster to call toLongFast and test (xl == (int) xl && yl == (int) yl). If there's no speed gain, let's exercise the 128-bit multiply for all non-big operands.

Reply:
> Is the int32 * int32 special case still warranted? It might be faster to call toLongFast and test (xl == (int) xl && yl == (int) yl).

With this diff on top of this PR:

```
--- a/src/main/java/net/starlark/java/eval/StarlarkInt.java
+++ b/src/main/java/net/starlark/java/eval/StarlarkInt.java
@@ -508,16 +508,14 @@ public abstract class StarlarkInt implements StarlarkValue, Comparable<StarlarkI

   /** Returns x * y. */
   public static StarlarkInt multiply(StarlarkInt x, StarlarkInt y) {
-    if (x instanceof Int32 && y instanceof Int32) {
-      long xl = ((Int32) x).v;
-      long yl = ((Int32) y).v;
-      return StarlarkInt.of(xl * yl);
-    }
-
     try {
       long xl = x.toLongFast();
       long yl = y.toLongFast();

+      if ((int) xl == xl && (int) yl == yl) {
+        return StarlarkInt.of(xl * yl);
+      }
+
       // see Hacker's Delight, 8-2

       long u0 = xl & 0xFFFFFFFFL;
```

For test `1 * 2`, result is 1% slowdown:

```
A: n=100 mean=3.886 std=0.182 se=0.018 min=3.563 med=3.852
B: n=100 mean=3.930 std=0.179 se=0.017 min=3.630 med=3.881
B/A: 1.011 0.998..1.025 (95% conf)
```

A: this PR
B: this PR + patch excluding special case

So I'd keep special case `instanceof Int32`.

Reply:
> I notice that the you (and the JDK) use a logical shift (t >>> 32) whereas HD uses an arithmetic shift (t >> 32). Why is that safe? I can't prove that t is nonnegative, though I can't yet find a bug.

I admit I don't completely understand this algorithm. But I change >>> to >>, tests start to fail.

Reply:
> So I'd keep special case instanceof Int32.

Thanks. I suppose virtual calls are more expensive than instanceof tests on final classes.

> I admit I don't completely understand this algorithm. But I change >>> to >>, tests start to fail.

Ok, well at least we know the tests cover it.